### PR TITLE
Add pytree support to pjit's with_sharding_constraint

### PR DESCRIPTION
**Pull request by skye**
Manually tested on a Cloud TPU VM.

### CODE REVIEW

**Comment by skye** [comment]
cc @levskaya

File: tests/pjit_test.py [hunk 1]
```diff
@@ -166,6 +166,30 @@ def f(x):
     # Annotation from pjit
     self.assertIn("sharding={replicated}", hlo.as_hlo_text())
 
+  @with_mesh([('x', 2), ('y', 1)])
+  def testShardingConstraintPyTree(self):
+    @partial(pjit, in_axis_resources=None, out_axis_resources=None)
+    def f(x):
+      x = with_sharding_constraint(x, [P('x', 'y'), P('y', 'x')])
+      x[0]["a"] *= 2
```
Comment (mattjj) [inline tests/pjit_test.py:174]:
```suggestion
      x = x.copy()
      x[0]["a"] *= 2
```

Otherwise there's a side-effect on the caller's dict, right?

**Review by mattjj — APPROVED** [review]
LGTM!

File: tests/pjit_test.py (hunk 1 above)
Comment (skye) [reply tests/pjit_test.py:174]: Haha I'm breaking the cardinal rule of JAX! Good catch.

**Comment by levskaya** [comment]
@skye this looks good, is there a reason it didn't go in?

**Comment by skye** [comment]
I'm not sure! Hopefully it just got lost in the copybara flow and I wasn't in the middle of fixing something I now forget about... I guess if the test passes it should be good? :)

**Pull request merged**